File: src/main/java/org/lanternpowered/server/util/option/Option.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.util.option;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

import com.google.gson.reflect.TypeToken;

import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

import org.checkerframework.checker.nullness.qual.Nullable;

public final class Option<V> {

    public static <V> Option<V> of(String id, TypeToken<V> typeToken, @Nullable V defaultValue) {
        return new Option<>(checkId(id), typeToken, defaultValue);
    }

    public static <V> Option<V> of(String id, TypeToken<V> typeToken) {
        return new Option<>(checkId(id), typeToken, null);
    }

    public static <V> Option<V> of(String id, Class<V> type) {
        return new Option<>(checkId(id), TypeToken.get(type), null);
    }

    public static <V> Option<V> of(String id, Class<V> type, @Nullable V defaultValue) {
        return new Option<>(checkId(id), TypeToken.get(type), defaultValue);
    }

    public static Optional<Option<?>> get(String id) {
        return Optional.ofNullable(options.get(checkNotNull(id, "id").toLowerCase(Locale.ENGLISH)));
    }

    private static final Map<String, Option<?>> options = new ConcurrentHashMap<>();

    private static String checkId(String id) {
        id = id.toLowerCase(Locale.ENGLISH);
        checkArgument(!options.containsKey(id), "The id %s is already used.", id);
        return id;
    }

    private final String id;
    private final TypeToken<V> typeToken;
    @Nullable private final V defaultValue;

    private Option(String id, TypeToken<V> typeToken, @Nullable V defaultValue) {
        this.typeToken = checkNotNull(typeToken, "typeToken");
        this.id = checkNotNull(id, "id").toLowerCase(Locale.ENGLISH);
        this.defaultValue = defaultValue;
        options.put(id, this);
    }

    /**
     * Gets the id of this option.
     *
     * @return The id
     */
    public String getId() {
        return this.id;
    }

    /**
     * Gets the {@link TypeToken} that represents the value type.
     *
     * @return The value type token
     */
    public TypeToken<V> getTypeToken() {
        return this.typeToken;
    }

    /**
     * Gets the default value of this option.
     *
     * @return The default value
     */
    public Optional<V> getDefaultValue() {
        return Optional.ofNullable(this.defaultValue);
    }
}

File: src/main/kotlin/org/lanternpowered/api/util/option/Option.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
@file:Suppress("UNCHECKED_CAST")

package org.lanternpowered.api.util.option

import org.lanternpowered.api.namespace.NamespacedKey
import org.lanternpowered.api.util.type.TypeToken
import kotlin.reflect.KClass

/**
 * Represents a option.
 *
 * @constructor Constructs and registers a new [Option]
 * @property defaultValue The default option value
 * @property type The type of option value
 * @param T The option map type this option targets
 * @param V The value type of this option
 */
data class Option<T : OptionMapType, V>(
        val key: NamespacedKey,
        val type: TypeToken<V>,
        val defaultValue: V,
        val mapType: Class<T>
) {

    constructor(key: NamespacedKey, type: TypeToken<V>, defaultValue: V, mapType: KClass<T>) :
            this(key, type, defaultValue, mapType.java)

    init {
        val registry = OptionMapRegistry.of(this.mapType)
        // Each key per map type should be unique
        check(this.key !in registry) {
            "There is already a option with id '$key' registered for map type '${mapType.name}'." }
        // Register the key
        registry.byKey[this.key] = this
    }
}
